ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,363 @@
+execution and possible errors that might surface.
+
+At the end, when the build-pod is completed successfully _(5)_, the CLI shows the meta information
+about the container image built, and instructions on how to pull and use the image.
+
+#### Watching the Build-Pod
+
+As mentioned before, Shipwright builds are based on a sequence of containers being executed on a
+single pod, the CLI needs to watch _(5)_ over the state changes, and react as soon as it happens. For
+example, the build-pod can be deleted, fail immediately, fail during execution, and such. Moreover,
+as we can observe on the sequence diagram, it also needs to wait for specific conditions before
+streaming data.
+
+During research and development, this requirement was translated into a [watch][ClientGoWatch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][EventLoopExample].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][EventLoopUsage].
+
+### Upload Mechanism
+
+This EP describes the direct data streaming to the Kubernetes cluster, directly to the POD running
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the `/workspace/source` mount point for all containers in the build-pod.
+Shipwright Build Controller also adds it's own helper to ["git clone" the
+repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+In the `BuildRun` resource, the Build Controller will inspect for the following annotations:
+
+- `build.shipwright.io/upload.wait`: boolean attribute, only `true` or `false` accepted;
+- `build.shipwright.io/upload.wait.timeout`: sets the timeout;
```

Comment:
I was expecting the upload method (push to image registry vs. kubectl cp, the openshift tar to an endpoint style, etc) to be one of the annotations @otaviof, even if we had only 1 choice initially

that for me at least also speaks to the first class api field vs. annotation subject, for among other things, some of the uncertainty with respect which of the upload implementation options we want to provide, whether any of them are a silver bullet, or if they each have enough cons to go with their pros that users of this feature will be somewhat spread out across the choices, is key factor at least for me for going down the annotation route for expressing choice, vs. first class API fields

I don't remember if this particular item was voiced, but I left the community meeting discussion between you, me, @HeavyWombat , and @imjasonh with this thought .... for references, this was a meeting where several folks missed the meeting with various conflicts, and our execution of the meeting was so haphazard I don't think we had a tracking issue).  This was the discussion where @HeavyWombat indicated there was push back from some usability folks for the code engine offering around using push to registry as a choice for the upload mechanism.

Reply:
> ... key factor at least for me for going down the annotation route for expressing choice, vs. first class API fields

Thanks! The explanation of items portrayed on the annotations must be equivalent to a statically typed fields in `.spec`, makes it clear for me! Indeed, makes sense.

I'm changing the annotation names to convey the `buildrun` wording, and introducing a new annotation to express the type of data-upload mechanism. So, initially only `buildrun.shipwright.io/upload.type: "direct"` would be available, but then we would as well have `bundle-image` later on.

Reply:
ok cool ... I like the pattern of all the annotations starting with the same prefix 

I do wonder if rather `upload` if it should be `localSource` ....we ultimately might have other "local source" tuning params beside things specific to the upload  ..... language detection a la oc new-app comes to mind

you up for changing to that @otaviof ?